aiidalab_ispg/spectrum_analysis.py

```diff
@@ -224,3 +241,389 @@ def _observe_disabled(self, change):
+            ),
+            "right",
+        )
+
+        return figure
+
+    # **********************************************************************
+
+    # **********************************************************************
+    # Observing changes in widget and spectrum data, update j plot for each change
+
+    @traitlets.observe("spectrum_data")
+    def _observe_spectrum_data(self, change):
+        """
+        Observe changes to the spectrum data and update the J plot accordingly.
+        Check that fluxdata overlaps with the spectrum data.
+
+        :param change: The change object containing the old and new values of the spectrum data.
+        :type change: dict
+        """
+        self.disabled = True
+        if change["new"] is None or len(change["new"]) == 0:
+            self.reset()
+            return
+
+        flux_min = min(self.flux_data[0])
+        spectrum_max = max(self.spectrum_data[0])
+
+        # Check end of spectrum data overlaps with flux data
+        if spectrum_max >= flux_min:
+            self._update_j_plot(
+                plot_type=self.flux_toggle.value, quantumY=self.yield_slider.value
+            )
+        else:
+            self.reset()
+            return
+
+        self.disabled = False
+
+    def _observe_flux_toggle(self, change):
+        """Redraw spectra when user changes flux via toggle
+
+        :param change: The change object containing the old and new values of the flux toggle.
+        :type change: dict
+        """
+        self._update_j_plot(plot_type=change.new, quantumY=self.yield_slider.value)
+
+    def handle_slider_change(self, change):
+        """Redraw spectra when user changes quantum yield via slider
+
+        :param change: The change object containing the old and new values of the yield slider.
+        :type change: dict
+        """
+        self._update_j_plot(plot_type=self.flux_toggle.value, quantumY=change.new)
+
+    # **********************************************************************
+
+    # **********************************************************************
+    def _update_j_plot(self, plot_type: str, quantumY):
+        """
+        Update the J plot based on the given plot type and quantum yield
+
+        :param plot_type: The flux of plot to generate. Can be "LOW", "MED", or "HIGH".
+        :type plot_type: str
+        :param quantumY: The quantum yield value to use in the calculation.
+        :type quantumY: float
+        :return: A tuple containing the J values and wavelengths used in the plot.
+        :rtype: tuple
+        """
+
+        if self.spectrum_data is None:
+            return
+        if plot_type == "LOW":
+            j_values = self.calculation(1, quantum_yield=quantumY)
+            wavelengths = self.flux_data[0]
+            self.plot_line(wavelengths, j_values, label="label")
+            self.add_log_axis(wavelengths, 1, label="log")
```

Comment:
There is a lot of repetition here. I would propose the `self.flux_data` to be a dictionary, and then you can use the `plot_type` as a key, and just pass it down to the `plot_line` and `add_log_axis` functions. 
Also the strings "LOW", "HIGH" and "MED" are arbitrarily defined in multiple places which is a bit error prone. It would be nicer to use the Enum type for this, in the same way as I use it in spectrum.py for other such variables. 

However, I would propose to do this refactoring in a separate followup PR.